staging/src/k8s.io/apiserver/pkg/storage/etcd3/etcd3retry/retry_etcdclient.go

```diff
@@ -0,0 +1,209 @@
+// match 'opts.ResourceVersion' according 'opts.ResourceVersionMatch'.
+func (c *retryClient) GetToList(ctx context.Context, key string, opts storage.ListOptions, listObj runtime.Object) error {
+	return onError(defaultRetry, alwaysRetry, func() error {
+		return c.Interface.GetToList(ctx, key, opts, listObj)
+	})
+}
+
+// List unmarshalls jsons found at directory defined by key and opaque them
+// into *List api object (an object that satisfies runtime.IsList definition).
+// The returned contents may be delayed, but it is guaranteed that they will
+// match 'opts.ResourceVersion' according 'opts.ResourceVersionMatch'.
+func (c *retryClient) List(ctx context.Context, key string, opts storage.ListOptions, listObj runtime.Object) error {
+	return onError(defaultRetry, alwaysRetry, func() error {
+		return c.Interface.List(ctx, key, opts, listObj)
+	})
+}
+
+// GuaranteedUpdate keeps calling 'tryUpdate()' to update key 'key' (of type 'ptrToType')
+// retrying the update until success if there is index conflict.
+// Note that object passed to tryUpdate may change across invocations of tryUpdate() if
+// other writers are simultaneously updating it, so tryUpdate() needs to take into account
+// the current contents of the object when deciding how the update object should look.
+// If the key doesn't exist, it will return NotFound storage error if ignoreNotFound=false
+// or zero value in 'ptrToType' parameter otherwise.
+// If the object to update has the same value as previous, it won't do any update
+// but will return the object in 'ptrToType' parameter.
+// If 'suggestion' can contain zero or one element - in such case this can be used as
+// a suggestion about the current version of the object to avoid read operation from
+// storage to get it.
+//
+// Example:
+//
+// s := /* implementation of Interface */
+// err := s.GuaranteedUpdate(
+//     "myKey", &MyType{}, true,
+//     func(input runtime.Object, res ResponseMeta) (runtime.Object, *uint64, error) {
+//       // Before each invocation of the user defined function, "input" is reset to
+//       // current contents for "myKey" in database.
+//       curr := input.(*MyType)  // Guaranteed to succeed.
+//
+//       // Make the modification
+//       curr.Counter++
+//
+//       // Return the modified object - return an error to stop iterating. Return
+//       // a uint64 to alter the TTL on the object, or nil to keep it the same value.
+//       return cur, nil, nil
+//    },
+// )
+func (c *retryClient) GuaranteedUpdate(
+	ctx context.Context, key string, ptrToType runtime.Object, ignoreNotFound bool,
+	precondtions *storage.Preconditions, tryUpdate storage.UpdateFunc, suggestion ...runtime.Object) error {
+	return onError(defaultRetry, isRetriableEtcdError, func() error {
+		return c.Interface.GuaranteedUpdate(ctx, key, ptrToType, ignoreNotFound, precondtions, tryUpdate, suggestion...)
+	})
+}
+
+func isRetriableEtcdError(err error) bool {
+	if err == nil {
```

Comment:
if the err could be read as `rpctypes.Error(err)` (https://github.com/etcd-io/etcd/blob/master/etcdserver/api/v3rpc/rpctypes/error.go#L230)  then according to https://github.com/etcd-io/etcd/blob/master/clientv3/retry.go#L53 all immutable errors that map to `codes.Unavailable` are safe to retry